### ISSUE TITLE
Firefox 127 not playing audio out Channels assigned in Audio MIDI setup

### ISSUE THREAD

**Issue by ChunMinChang** [mozilla/cubeb-coreaudio-rs#232]
This issue is mirrored from [BMO 1904734](https://bugzilla.mozilla.org/show_bug.cgi?id=1904734), as detailed in [comments #6](https://bugzilla.mozilla.org/show_bug.cgi?id=1904734#c6) and [#7](https://bugzilla.mozilla.org/show_bug.cgi?id=1904734#c6).

On my [Cubilux Surround Sound Card (CA7)](https://www.cubilux.com/products/7-1-5-1surround-sound-crad), the channel layout is set to `[BackLeft, BackRight, FrontCenter, LowFrequency, FrontLeft, FrontRight]`. However, audio only comes out of the first two channels when tested in Firefox and with [`test_tone`](https://github.com/mozilla/cubeb-coreaudio-rs/blob/8bce3b333a920999055397a397e59c2b81a93b9a/src/backend/tests/manual.rs#L773). With a 2-channel input and a 6-channel output, a [Mixer](https://github.com/mozilla/cubeb-coreaudio-rs/blob/8bce3b333a920999055397a397e59c2b81a93b9a/src/backend/mixer.rs#L166) should be created to upmix the audio.

I am not sure what [`kAudioUnitProperty_AudioChannelLayout` setting](https://github.com/mozilla/cubeb-coreaudio-rs/blob/8bce3b333a920999055397a397e59c2b81a93b9a/src/backend/mod.rs#L3877-L3899)  is supposed to do, but it doesn't seem to change my Cubilux device setting. Instead, it sets [`maybe_need_mixer`](https://github.com/mozilla/cubeb-coreaudio-rs/blob/8bce3b333a920999055397a397e59c2b81a93b9a/src/backend/mod.rs#L3901) to `false`, preventing the mixer from being created.

Commenting out the `kAudioUnitProperty_AudioChannelLayout` setting doesn't solve the problem, as the [`output_dev_desc`](https://github.com/mozilla/cubeb-coreaudio-rs/blob/8bce3b333a920999055397a397e59c2b81a93b9a/src/backend/mod.rs#L3929)'s [`mChannelsPerFrame`](https://github.com/mozilla/cubeb-coreaudio-rs/blob/8bce3b333a920999055397a397e59c2b81a93b9a/src/backend/mod.rs#L3918) still reports `2`. This leads to creating a 2-channel output mixer, even though the [`device_layout`](https://github.com/mozilla/cubeb-coreaudio-rs/blob/8bce3b333a920999055397a397e59c2b81a93b9a/src/backend/mod.rs#L3857) correctly reflects the channel layout.

The channels of `output_dev_desc` are [defined](https://github.com/mozilla/cubeb-coreaudio-rs/blob/8bce3b333a920999055397a397e59c2b81a93b9a/src/backend/mod.rs#L3820) by `output_hw_desc.mChannelsPerFrame`, which is expected to reflect the hardware information. However, it also reports `2` in this case. The `output_hw_desc` is [queried](https://github.com/mozilla/cubeb-coreaudio-rs/blob/8bce3b333a920999055397a397e59c2b81a93b9a/src/backend/mod.rs#L3782-L3791) from the `kAudioUnitProperty_StreamFormat property`. Changing the query scope from [`kAudioUnitScope_Input`](https://github.com/mozilla/cubeb-coreaudio-rs/blob/8bce3b333a920999055397a397e59c2b81a93b9a/src/backend/mod.rs#L3787) to `kAudioUnitScope_Output` makes `output_hw_desc` report the expected `mChannelsPerFrame` value of `6`, resulting in an `output_dev_desc` setting that allows the creation of a 6-channel output mixer.

In summary, commenting out the `kAudioUnitProperty_AudioChannelLayout` setting and changing the query scope for `kAudioUnitProperty_StreamFormat` fixes the issue. These changes were introduced in PR #220 and PR #222. However, there were no comments or explanations detailing the issues they were meant to address. Without this context, I am unsure if reverting these changes is the right approach or if it might lead to other regressions.

@Pehrsons, @padenot, do you remember what those settings do? I'd like to leave some discussion here or on the bug for future reference. Feel free to comment with your thoughts about the right approach.

**Comment by ChunMinChang** [comment]
According to [BMO 1904734 comment #10](https://bugzilla.mozilla.org/show_bug.cgi?id=1904734#c10), it seems [kAudioUnitProperty_AudioChannelLayout setting](https://github.com/mozilla/cubeb-coreaudio-rs/blob/8bce3b333a920999055397a397e59c2b81a93b9a/src/backend/mod.rs#L3877-L3899) should be applied if only `FrontLeft` and `FrontRight` are meaningful in the channel layout, e.g, `[ .., Discrete, ...., FrontLeft, ..., FrontRight, ..., Silence, ...]`. We could probably add a `device_layout` inspection in the if conditions for `maybe_need_mixer`.

**Comment by ChunMinChang** [comment]
Currently, #233 partially addresses the issue. Other commits/patches are required to fix the problem.

**Comment by padenot** [comment]
When playing stereo, audio should come out of the stereo pair, is that not what you're expecting?

**Comment by Pehrsons** [comment]
Testing with Blackhole16ch, having set Front Left to ch7 and Front Right to ch8 and playing stereo audio with test_tone I get the left channel playing through ch7 and the right through ch8 as expected. Mono got downmixed (half volume or so) to ch7 and ch8. With #233 I got Mono at full volume to ch3 instead (Center).

I guess this is the difference in mixer behavior when outputting 8ch (with a Center channel) vs 2ch (without Center).

**Comment by ChunMinChang** [comment]
Here is the brief of the offline discussion:
- By [kAudioUnitProperty_AudioChannelLayout setting](https://github.com/mozilla/cubeb-coreaudio-rs/blob/8bce3b333a920999055397a397e59c2b81a93b9a/src/backend/mod.rs#L3877-L3899) introduced in #222, for **2-channel** audio input, the output channels follows the *stereo* configuration, rather than other surround configurations: 
![Screen Shot 2024-08-21 at 10 47 46 AM](https://github.com/user-attachments/assets/59067e58-f4ed-49c4-8948-522d2f10ac2d)
- For **non** 2-channel audio input, the output channels follow the surround configurations
- The above decision follows the *Apple Music* setting, which always apply the *stereo* configuration
- Safari and Chrome always apply surround configurations instead of *stereo* configurations even though the input audio is 2-channel.

**Comment by ChunMinChang** [comment]
I am going to close this since #233 solves the channel count issue and leads to addressing the needs above.
